kratos/utilities/point_locator.cpp

```diff
@@ -0,0 +1,148 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:		 BSD License
+//					 Kratos default license: kratos/license.txt
+//
+//  Main authors:    Philipp Bucher (based on work of Pablo Becker)
+//
+
+
+// System includes
+
+
+// External includes
+
+
+// Project includes
+#include "point_locator.h"
+
+
+namespace Kratos
+{
+    int PointLocator::FindNode(const Point& rThePoint,
+                               const double DistanceThreshold) const
+    {
+        int found_node_id = -1; // if no node is found this will be returned
+        bool is_close_enough = false;
+
+        int global_nodes_found = 0;
+
+        // note that this cannot be omp bcs breaking is not allowed in omp
+        for (auto& r_node : mrModelPart.GetCommunicator().LocalMesh().Nodes())
+        {
+            is_close_enough = NodeIsCloseEnough(r_node, rThePoint, DistanceThreshold);
+            if (is_close_enough)
+            {
+                global_nodes_found = 1;
+                found_node_id = r_node.Id();
+                break;
+            }
+        }
+
+        CheckResults("Node", rThePoint, global_nodes_found);
+
+        return found_node_id;
+    }
+
+    int PointLocator::FindElement(const Point& rThePoint,
+                                  Vector& rShapeFunctionValues) const
+    {
+        int found_element_id = -1; // if no element is found this will be returned
+        const auto& r_elements = mrModelPart.GetCommunicator().LocalMesh().Elements();
+        FindObject(r_elements, "Element",
+                   rThePoint, found_element_id,
+                   rShapeFunctionValues);
+        return found_element_id;
+    }
+
+    int PointLocator::FindCondition(const Point& rThePoint,
+                                    Vector& rShapeFunctionValues) const
+    {
+        int found_condition_id = -1; // if no condition is found this will be returned
+        const auto& r_conditions = mrModelPart.GetCommunicator().LocalMesh().Conditions();
+        FindObject(r_conditions, "Condition",
+                   rThePoint, found_condition_id,
+                   rShapeFunctionValues);
+        return found_condition_id;
+    }
+
+    template<typename TObjectType>
+    void PointLocator::FindObject(const TObjectType& rObjects,
+                                  const std::string& rObjectName,
+                                  const Point& rThePoint,
+                                  int& rObjectId,
+                                  Vector& rShapeFunctionValues) const
+    {
+
+        const int domain_size = mrModelPart.GetProcessInfo()[DOMAIN_SIZE];
+
+        const auto& r_geom = rObjects.begin()->GetGeometry();
+
+        KRATOS_ERROR_IF_NOT(static_cast<std::size_t>(domain_size) == r_geom.WorkingSpaceDimension())
```

Comment:
As a note :  This check is also happening inside the function "IsInside"  (in the geometry base class). This need not be done here.

Reply:
this should be in debug, but not in release

Reply:
I really think that this should be in release and not only in debug

Otherwise I am silently modifying the bahavior, which will be very hard to debug!

Plus the cost of this is small compared to the computation of this clss

Reply:
i am still convinced that these checks belong to debug and not release. 

alsp what about @adityaghantasala comment?